### ISSUE TITLE
Documentation shows None default instead of actual op attr default

### ISSUE THREAD

**Issue by Antonio95** [tensorflow/tensorflow#5667]
This is not a bug but a small mistake in the API:
The [documentation ](https://www.tensorflow.org/versions/r0.11/api_docs/python/state_ops.html#assign) says "use_locking" defaults to _None_ in **tf.assign**.
That parameter is a boolean, so _None_ would not make a lot of sense. In any case, the source code makes it clear the default value is actually _False_ (file: tensorflow/tensorflow/python/ops/variables.py)

I do not know if this is the case with other boolean default arguments in other functions, but I wanted to point it out.

**Comment by girving** [comment]
@Antonio95 Yep, this is a problem with our doc generator.  The place that correctly says `False` is a different thing: it's a convenience method of `Variable` that calls down to the underlying op.

This would be great to fix, but we may not get to it soon.  Contributions would be very welcome!

More background: this documentation is for a Python wrapper automatically generated based on the C++ op definition.  The Python wrappers use `None` as defaults.  Is there a good reason for that, @josh11b (I vaguely recall having the discussion but don't remember the result)?  In any case, once the Python wrapper has a `None` default, the generated docs do too since they pull the defaults out of the Python function definition.

**Comment by Antonio95** [comment]
I'd be very happy to help (I'm writing my final degree project on TF (python API)), if you tell me how to get started.

**Comment by girving** [comment]
Thank you!

How to proceed depends on whether @josh11b and @vrv think we should fix the Python wrapper generator or the doc generator.  Fixing the Python wrapper generator is more natural, but there may be a subtle reason for the `None`s that I do not recall.

**Comment by vrv** [comment]
I was under the impression that there was an overhaul to our doc generation / website coming and that's why this wasn't being fixed ...  @martinwicke ?

**Comment by Antonio95** [comment]
As soon as you make a decision I'll be happy to try and learn how to solve the issue

**Comment by josh11b** [comment]
To re-iterate what girving@ said:
* The documentation you link to does not come from tensorflow/python/ops/variables.py, but from the generated code (gen_state_ops.py, imported by tensorflow/python/ops/state_ops.py).
* I'm a bit foggy on the specifics of why we did generated the code the way we did (the documentation just comes from the generated code), but I can hazard a few guesses:
 - It avoid duplicating the places where we specify the default, reducing friction if we ever want to change the default (especially since sometimes people copy-paste the generated code as a starting point for a wrapper around it).
 - We'd have to write code to render the default as legal Python in the generated code, which was extra code we didn't really need for anything else.
 - There is a gotcha if the default value is considered mutable by Python, see e.g.: http://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html

Instead we put the defaults in the doc string (search for the string "Defaults to").  We talked about various fixes (for example changing the rendering from "arg=None" to something else that indicates it is optional without specifying what the default is, perhaps "[arg]"), but no one solution made everybody happy so we left it as-is.  Another (tricky) possible solution: get the default out of the doc string, but that seems like it might be a bit fragile (e.g. could be disturbed by word wrapping, or a dot inside the default value).
